wasm/derive/src/entrypoint.rs

```diff
@@ -0,0 +1,160 @@
+
+mod kw {
+    syn::custom_keyword!(params);
+
+    pub mod param_types {
+        syn::custom_keyword!(authority);
+        syn::custom_keyword!(triggering_event);
+    }
+}
+
+/// Enum representing possible attributes for [`entrypoint`] macro
+enum Attr {
+    Params(ParamsAttr),
+    Empty,
+}
+
+impl syn::parse::Parse for Attr {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        if input.is_empty() {
+            return Ok(Attr::Empty);
+        }
+
+        Ok(Attr::Params(input.parse()?))
+    }
+}
+
+/// Attribute with expected parameters for smart contract entrypoint function
+struct ParamsAttr {
+    _params_kw: kw::params,
+    _equal: syn::token::Eq,
+    params: Params,
+}
+
+impl syn::parse::Parse for ParamsAttr {
+    fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
+        let params_kw = input.parse()?;
+        let equal = input.parse()?;
+        let params_str: syn::LitStr = input.parse()?;
+        let params = syn::parse_str(&params_str.value())?;
+        Ok(ParamsAttr {
+            _params_kw: params_kw,
+            _equal: equal,
+            params,
+        })
+    }
+}
+
+/// Collection of types of parameters that smart contract entrypoint function is expecting
+struct Params {
+    _bracket_token: syn::token::Bracket,
+    types: Punctuated<ParamType, syn::token::Comma>,
+}
+
+impl syn::parse::Parse for Params {
+    fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
+        let content;
+        let bracket_token = syn::bracketed!(content in input);
+
+        Ok(Params {
+            _bracket_token: bracket_token,
+            types: content.parse_terminated(ParamType::parse)?,
+        })
+    }
+}
+
+/// Type of smart contract entrypoint function parameter.
+///
+/// *Type* here means not just *Rust* type but also a purpose of a parameter.
+/// So that it uses [`Authority`](ParamType::Authority) instead of `account::Id`.
+enum ParamType {
+    Authority,
+    TriggeringEvent,
+}
+
+impl syn::parse::Parse for ParamType {
+    fn parse(input: syn::parse::ParseStream) -> syn::parse::Result<Self> {
+        if input.parse::<kw::param_types::authority>().is_ok() {
+            Ok(ParamType::Authority)
+        } else if input.parse::<kw::param_types::triggering_event>().is_ok() {
+            Ok(ParamType::TriggeringEvent)
+        } else {
+            Err(input.error("expected `authority` or `triggering_event`"))
+        }
+    }
+}
+
+/// [`entrypoint`](crate::entrypoint()) macro implementation
+pub fn impl_entrypoint(attr: TokenStream, item: TokenStream) -> TokenStream {
+    let syn::ItemFn {
+        attrs,
+        vis,
+        sig,
+        mut block,
+    } = parse_macro_input!(item);
+
+    assert!(
+        syn::ReturnType::Default == sig.output,
+        "Exported function must not have a return type"
+    );
+
+    let args = match syn::parse_macro_input!(attr as Attr) {
```

Comment:
I don't think so. We can't check types of a function signature. Using string comparision instead of type checking is bad. I think, the approach with attributes is pretty ok

Reply:
> I think, the approach with attributes is pretty ok
 
 Yes, sure